Инкремент 5

Fairness2 opened the pull request:
Добавлена поддержка конфигурации из переменных окружения

Fairness2 committed:
iter1 | Добавлен сервер для работы с метриками и обработчики запросов

Был добавлен функционал для обработки HTTP запросов, добавления метрик. Также созданы вспомогательные функции для работы с метриками: установки, получения и ошибок HTTP.

Измененные файлы:
cmd/server/handlers/handlemetric.go,
cmd/server/main.go,
cmd/server/handlers/getmetric.go,
internal/metrics/MemStorage.go,
go.mod,
internal/metrics/Metrics.go,
.gitignore,
cmd/skill/main.go.

Fairness2 committed:
iter1.1 | Изменён стиль именования и форматирования сообщений в handlemetric.go

В соответствии с рекомендациями по стилю кода все функции и сообщения об ошибках были приведены к единообразному стилю

Измененный файл:
cmd/server/handlers/handlemetric.go

Fairness2 committed:
iter2 | Реализован сбор и отправка метрик серверу, добавлены тесты и доработан gitignore

Была внедрена функциональность сбора системных метрик и отправки их на сервер. Добавлены юнит тесты для новогой функциональности. В .gitignore добавлены пути до исполняемых файлов сервера и агента.

Измененные файлы:
.gitignore
cmd/agent/main.go
internal/metrics/MemStorage.go
cmd/skill/main_test.go
go.mod
internal/metrics/Metrics.go
cmd/server/handlers/handlemetric_test.go
internal/metrics/MemStorage_test.go
cmd/agent/env/env.go
cmd/agent/collector/sender/sender_test.go
cmd/server/handlers/handlemetric.go
cmd/agent/collector/main.go
cmd/agent/collector/collection/collection.go
cmd/agent/collector/sender/sender.go

Fairness2 committed:
iter2 | Обновлены параметры работы агента

В коде агента внесены изменения, связанные с ожиданием и работой с сервером. Удалена возможность прекращения работы агента командой из stdin, вместо этого добавлено ожидание завершения работы в течение 30 секунд.
Также было изменено название переменной ServerUrl на ServerURL

Измененные файлы:
cmd/agent/main.go
cmd/agent/env/env.go
cmd/agent/collector/sender/sender_test.go
cmd/agent/collector/sender/sender.go

Fairness2 committed:
iter2.2 | Добавлен новый файл go.sum

Был добавлен новый файл go.sum, который содержит хеши зависимостей проекта.

Измененный файл:
go.sum

Fairness2 committed:
iter2.3 | Исправлены наименования переменных и добавлен закрытие ответа в sender.go

Были произведены изменения в двух файлах: Metrics.go и sender.go. В Metrics.go было исправлено наименование переменной в методе ToString(). В sender.go исправлено наименование переменной и добавлено закрытие tела ответа.

Измененные файлы:
internal/metrics/Metrics.go
cmd/agent/collector/sender/sender.go

Fairness2 committed:
iter2.4 | Исправлено наименование переменной и добавлено закрытие тела ответа в handlemetric_test.go

Измененный файл:
cmd/server/handlers/handlemetric_test.go

Fairness2 committed:
iter3 | Переписан код для использования библиотек resty и chi. Добавлены обработчики для получения метрик

Часть обработчиков была перенесена в отдельные модули. Был проведен рефакторинг кода и тестов, для улучшения переиспользуемости кода и снижения сложности.
Также были добавлены новые вспомогательные функции, в том числе функции для обработки ошибок HTTP.

Измененные файлы:
cmd/server/handlers/getmetric.go
cmd/agent/main.go
internal/helpers/helpers.go
cmd/server/main.go
internal/helpers/helpers_test.go
internal/metrics/MemStorage.go
cmd/skill/main_test.go
go.mod
go.sum
cmd/server/handlers/handleMetric/handlemetric_test.go
internal/metrics/MemStorage_test.go
cmd/agent/collector/sender/sender_test.go

Fairness2 committed:
iter3.1 | Исправлены названия пакетов для соответствия Go conventions

Все названия пакетов были приведены к нижнему регистру без пробелов и спецсимволов, что соответствует рекомендованным соглашениям Go.

Измененные файлы:
cmd/server/handlers/handlemetric/handlemetric.go
cmd/server/main.go
cmd/server/handlers/getmetrics/handler.go
cmd/server/handlers/handlemetric/handlemetric_test.go
cmd/server/handlers/getmetric/handler.go

Fairness2 committed:
iter3.2 | Добавлены unit-тесты для обработчиков getmetric и getmetrics

Были созданы unit-тесты для обработчиков метрик getmetric и getmetrics. Тесты проверяют разбор URL на тип и имя метрики, а также обработку метрик. В ходе добавления тестов были внесены незначительные изменения в MemStorage и обработчиках.

Измененные файлы:
cmd/server/handlers/getmetric/handler_test.go
cmd/server/handlers/getmetrics/handler_test.go
cmd/server/handlers/getmetrics/handler.go
internal/metrics/MemStorage.go
cmd/server/handlers/getmetric/handler.go

Fairness2 committed:
iter3.3 | Добавлен комментарий к интерфейсу Storage в MemStorage.go

Измененный файл:
internal/metrics/MemStorage.go

Fairness2 committed:
iter3.3 | (((

Измененный файл:
cmd/server/handlers/getmetrics/handler_test.go

Fairness2 committed:
iter3.4 | (((

Измененный файл:
cmd/server/handlers/getmetrics/handler_test.go

Fairness2 committed:
iter4 | Конфигурация севера и агента через параметры командной строки

Изменены файлы:
cmd/agent/main.go,
cmd/server/main.go,
cmd/skill/flags.go,
cmd/server/config/config.go,
cmd/agent/collector/main.go,
cmd/agent/config/env.go,
.gitignore,
cmd/agent/collector/sender/sender_test.go,
cmd/agent/collector/sender/sender.go,
cmd/skill/main.go.

Fairness2 committed:
iter5 | Добавлена поддержка конфигурации из переменных окружения

В код внесены изменения, позволяющие получать параметры конфигурации из переменных окружения. Реализована обратная совместимость с конфигурацией из аргументов командной строки: в случае конфликта приоритет отдается переменной окружения.

Список измененных файлов:
cmd/skill/flags.go
go.mod
cmd/server/config/config.go
go.sum
cmd/agent/config/env.go

vokinneberg commented on cmd/agent/config/env.go:
```diff
@@ -0,0 +1,94 @@
+var ReportInterval = time.Duration(10) * time.Second
+
+// ServerURL Url сервера получателя метрик
+var ServerURL = "http://localhost:8080"
```
 Значения по-умолчанию можно объявить как константы. Не обязательно использовать `time.Duration(10)`, так как умножение `2 * time.Second`  так же даст тип `Duration`.

```suggestion
const (
	// PollInterval Интервал между сборкой данных
	PollInterval = 2 * time.Second

	// ReportInterval Интервал между отправкой данных
	ReportInterval = 10 * time.Second

	// ServerURL Url сервера получателя метрик
	ServerURL = "http://localhost:8080"
)
```

Fairness2 replied:
В константах убрал приведение к Duration

vokinneberg commented on cmd/agent/config/env.go:
```diff
@@ -0,0 +1,94 @@
+var ServerURL = "http://localhost:8080"
+
+// Parse инициализирует новую консольную конфигурацию, обрабатывает аргументы командной строки
+func Parse() {
```
Мне кажется, что использование глобальных переменных, которые устанавливаются после парсинга параметров командной строки и переменных окружения, слишком сложное. Предлагаю вариант, при котором метод `Parse(...)` возвращает инициализированную структуру конфига, а глобальные константы используются как значения по умолчанию. Логично, если параметры командной строки будут переопределять значения переменных окружения.

```go
type Config struct {
		PollInterval   int    `env:"POLL_INTERVAL" envDefault:"-2"`
		ReportInterval int    `env:"REPORT_INTERVAL" envDefault:"-2"`
		ServerURL      string `env:"ADDRESS"`
}

func Parse() *Config {
	...
}
```

Fairness2 replied:
Глобальные константы используются теперь для значений по умолчанию. Сам конфиг заполняется в структуру, которая заполняется по правилам окружение>командная строка>значения по умолчанию

vokinneberg commented on cmd/agent/config/env.go:
```diff
@@ -0,0 +1,94 @@
+	err := env.Parse(&cnf)
+	// Если ошибка, то считаем, что вывести конфигурацию из окружения не удалось
+	if err != nil {
+		log.Print(err)
```
Я бы рекомендовал вернуть ошибку в вызывающий код и там принять решение о том, как ее обрабатывать.

Fairness2 replied:
Теперь ошибка возвращается на верхний уровень и решения принимаются там

vokinneberg commented on cmd/agent/config/env.go:
```diff
@@ -0,0 +1,94 @@
+		ReportInterval = time.Duration(cnf.ReportInterval) * time.Second
+	}
+	if cnf.ServerURL != "" {
+		_ = setServerURL(cnf.ServerURL)
```
Не стоит игнорировать ошибку. Я бы рекомендовал обработать ошибку и вернуть ее в вызывающий код и там принять решение о том, как ее обрабатывать.

```suggestion
		if err := setServerURL(cnf.ServerURL); err != nil {
		        return fmt.Errof("set server URL: %v", err)
		}
```

Fairness2 replied:
Теперь ошибка не игнорируется

vokinneberg commented on cmd/agent/config/env.go:
```diff
@@ -0,0 +1,94 @@
+		if err != nil {
+			return errors.New("invalid frequency of metrics collection: " + err.Error())
+		}
+		PollInterval = time.Duration(val) * time.Second
```
Можно упростить выражение.

```suggestion
		PollInterval = val * time.Second
```

Fairness2 replied:
Не получится, компилятор жалуется на то, что не может спарсить инт в продолжительность 
![image](https://github.com/user-attachments/assets/ccb28395-bfe2-4b63-aea1-6d297a8d74d0)

vokinneberg replied:
Действительно, это работает только с литералом `PollInterval = 10 * time.Second`. Моя шибка 👍

vokinneberg commented on cmd/agent/main.go:
```diff
@@ -1,3 +1,33 @@
+
+	time.Sleep(30 * time.Second)
+
+	/*for {
+		fmt.Println("Agent is running. Print C to finish agent")
+		var command string
+		_, err := fmt.Fscan(os.Stdin, &command)
+		if err != nil {
+			fmt.Println(err)
+		}
+		if command == "C" {
+			break
+		}
+	}*/
```
Не стоит оставлять закомментированный код, так как он засоряет кодовую базу. Если это необходимо, стоит добавить комментарий с объяснением, почему данный код нужен.

Fairness2 replied:
Он оставлен, так как является лушим вариантом, чем таймер на время работы выше, но без таймера не проходят тесты

vokinneberg commented on cmd/agent/main.go:
```diff
@@ -1,3 +1,33 @@
+	collector.StartCollect()                          // Запускаем сборку данных
+	_ = collector.StartSending(collection.Collection) // Запускаем отправку данных
+
+	time.Sleep(30 * time.Second)
```
Не совсем понимаю какое назначение у этого таймера? Если он нужен для того чтобы ожидать выполнения, сбора метрик и из отправки, то это таймер сработает через 30 секунд и приложение закончит работу. Здесь стоит использовать другие средства синхронизации - рассмотри возможность использования каналов или  waitGroup.

Fairness2 replied:
Тут поставил таймер, так как с ожидание ввода, что закомментирован ниже, не проходили тесты, так как сервер работал слишком долго и не завершал свою работу. Я бы хотел убрать таймер и расскоментировать код с завершением по вводу пользователя, но тесты не пройдут

vokinneberg commented on cmd/agent/collector/main.go:
```diff
@@ -0,0 +1,38 @@
+package collector
```
Вообще по конвенции Go файл `main.go` - это файл, который содержит точку входа в приложение - функцию main. Поэтому лучше во избежание путаницы выбрать для это файла другое имя.

Fairness2 replied:
переименовано

vokinneberg commented on cmd/agent/collector/main.go:
```diff
@@ -0,0 +1,38 @@
+// collectProcess continuously collects system metrics by reading memory stats,
+// collecting the stats using collection.Collection.Collect, and then sleeping
+// for the duration defined by config.PollInterval.
+func collectProcess() {
```
В этом методе не хватает завершения цикла. Попробуй сделать завершение при помощи `context`  cancellation.

Fairness2 replied:
Использован контекст для выхода

vokinneberg commented on cmd/agent/collector/collection/collection.go:
```diff
@@ -0,0 +1,158 @@
+}
+
+// UnlockRead Завершение чтения с collection и освобождение mutex
+func (c *CollectionType) UnlockRead() {
```
Что здесь все-таки предполагается вызывать? Если `Unlock(...)`, то использование `sync.RWNutex` не совсем обосновано, так как его методы нигде не используются. К тому же по логике реализации похоже, что все-таки здесь предполагается использовать  `RUnlock(...)`?

Fairness2 replied:
Сделан обычный мютекс, который блокируется при любом обновлении, а так как у нас при чтении сбрасывается каунтер, то блокируем и тогда

vokinneberg commented on cmd/agent/collector/collection/collection.go:
```diff
@@ -0,0 +1,158 @@
+)
+
+// init Инициализация коллекции метрик и связанного мьютекса
+func init() {
```
Не рекомендую использовать метод `init()` для инициализации коллекции. Думаю, стоит создавать коллекцию в момент инициализации агента, а mutex можно сделать частью самой коллекции.

`init()` функции могут быть полезны для статической инициализаци конфигурации на уровне пакета, например. Но не стоит их использовать для создания инстансов структур и т.д. так как это может затруднить и отладку и тестирование.

Fairness2 replied:
Метод init больше не используется для инициализации значений

vokinneberg commented on cmd/agent/collector/collection/collection.go:
```diff
@@ -0,0 +1,158 @@
+	c.PollCount = c.PollCount.Add(1)
+	c.RandomValue = metrics.Gauge(rand.Float64())
+
+	fmt.Println("End collecting metrics")
```
Я бы рекомендовал использовать логгер либо из пакета `log` либо из пакета `slog` вместо `fmt.Println(...)`.

Fairness2 replied:
Перевёл всё на лог

vokinneberg commented on cmd/agent/collector/main.go:
```diff
@@ -0,0 +1,38 @@
+// launching a goroutine to execute the periodicSender function.
+func StartSending(coll *collection.CollectionType) *sender.Client {
+	client := sender.New(coll)
+	fmt.Println("New sender client created")
```
Я бы рекомендовал использовать логгер либо из пакета `log` либо из пакета `slog` вместо `fmt.Println(...)`.

Fairness2 replied:
Перевёл всё на log

vokinneberg commented on cmd/agent/collector/sender/sender.go:
```diff
@@ -0,0 +1,99 @@
+}
+
+// StartSender Старт начала прослушивания и создание клиента
+func (c *Client) StartSender() {
```
Не совсем понятно назначение метода-обертки `StartSender(...)`. Мне кажется будет проще сделать метод `periodicSender()` публичным, и вызывать его напрямую. Так же не стоит вызывать горутину не явно внутри другого метода - это может ввести в заблуждение о природе работы метода.

Fairness2 replied:
Убрал дополнительные обёртки над стартами потоков, теперь они явно вызываются в main

vokinneberg commented on cmd/agent/collector/sender/sender.go:
```diff
@@ -0,0 +1,99 @@
+	fmt.Println("Sending metrics")
+	c.metricsCollection.LockRead()
+	defer c.metricsCollection.UnlockRead()
+	v := reflect.Indirect(reflect.ValueOf(c.metricsCollection))
```
Рефлексия это мощный и в то же время опасный механизм. В общем случае не стоит использовать рефлексию там, где можно без нее обойтись. Здесь можно попробовать использовать мапу `map[string]int64` внутри коллекции для сборка метрик и потом отправлять метрики используя ключ - значение, которые хранятся в мапе.

Fairness2 replied:
Убрал использование рефлексии, а метрики перевёл на использование мапы

vokinneberg commented on cmd/agent/collector/sender/sender_test.go:
```diff
@@ -0,0 +1,117 @@
+			expectedError: false,
+		},
+		{
+			name: "invalid http status",
```
Не стоит использовать одинаковые имена для именования тестов. Лучше дать им более информативные названия, описывающие, что делает тест.

Fairness2 replied:
Поправил названия тестов

vokinneberg commented on cmd/agent/collector/sender/sender_test.go:
```diff
@@ -0,0 +1,117 @@
+		expectedError bool
+	}{
+		{
+			name: "successful metric send",
```
Небольшая рекомендация по именованию тестов - Go автоматически заменяет пробелы на подчерки в названиях тестов. Поэтому я бы не рекомендовал использовать пробелы в названиях или сразу заменять из на подчерки - это позволит избежать проблем про поиске упавших тестов.

Fairness2 replied:
Заменил пробелы на нижнее подчёркивание

vokinneberg commented on cmd/agent/collector/sender/sender_test.go:
```diff
@@ -0,0 +1,117 @@
+			c := New(getMockCollection())
+			err := c.sendMetric(tc.metricType, tc.metricName, tc.metricValue)
+			if tc.expectedError {
+				assert.Error(t, err)
```
В общем-то в использовании сторонних assert  библиотек нет ничего неправильного, но разработчики Go не рекоменду использовать сторонние assert библиотеки, так как _это либо останавливает тест раньше времени (если вызовы используют t.Fatalf или panic), либо упускает интересную информацию о том, что тест сделал правильно. Это также заставляет assert библиотеку создавать совершенно новый подъязык вместо использования существующего языка программирования (самого Go)._ https://go.dev/wiki/TestComments

Fairness2 replied:
В дальнешем буду стараться использовать чистый го, по опыту других языков привычнее пользоваться асертами просто

vokinneberg commented on cmd/server/config/config.go:
```diff
@@ -0,0 +1,46 @@
+}
+
+// Parse инициализирует новую консольную конфигурацию, обрабатывает аргументы командной строки
+func Parse() {
```
Предлагаю тут использовать такую же схему, как для создания конфига агента при котором метод `Parse(...)` возвращает инициализированную структуру конфига, а глобальные константы используются как значения по умолчанию.

Fairness2 replied:
Глобальные константы используются теперь для значений по умолчанию. Сам конфиг заполняется в структуру, которая заполняется по правилам окружение>командная строка>значения по умолчанию

vokinneberg commented on cmd/server/main.go:
```diff
@@ -1,3 +1,36 @@
+func main() {
+	config.Parse()                // Заполняем конфигурацию сервера
+	if err := run(); err != nil { // Запускаем сервер
+		panic(err)
```
Не стоит использовать панику, так как она приводит к крашу приложения. Рекомендую вызвать `log.Fatalf(...)`, которая выведет сообщение об ошибке и завершит процесс при помощи вызова `os.Exit(1)`.

Fairness2 replied:
Переделал на log.Fatalf

vokinneberg commented on internal/metrics/MemStorage_test.go:
```diff
@@ -0,0 +1,140 @@
+	testCases := []struct {
+		name      string
+		mName     string
+		wantValue interface{}
```
Можно заменить тип `interface{}` на тип `any`. Это синонимы, но второй предпочтительнее именно из за своего имени.

Fairness2 replied:
Заменил interface{} на any

vokinneberg commented on internal/metrics/MemStorage.go:
```diff
@@ -0,0 +1,82 @@
+package metrics
+
+// Storage represents an interface for accessing and manipulating metrics storage.
+type Storage interface {
```
👍 за реализацию интерфейса. Я бы рекомендовал попробовать определить интерфейс не рядом с реализацией, а рядом с тем местом, где он используется. Это позволит следовать принципам   Dependency Inversion и interface Segregation.

Fairness2 replied:
Перенёс интерфейс и глобальную переменную в отдельный файл internal/metrics/storage.go, где они используются

vokinneberg commented on internal/metrics/MemStorage_test.go:
```diff
@@ -0,0 +1,140 @@
+	"testing"
+)
+
+func TestNewMemStorage(t *testing.T) {
```
👍 за реализацию теста хранилища.

vokinneberg commented on internal/metrics/MemStorage.go:
```diff
@@ -0,0 +1,82 @@
+var MeStore Storage
+
+// Инициализация хранилища
+func init() {
```
Опять же не рекомендую использовать метод `init()` для инициализации коллекции. Думаю, стоит создавать коллекцию в момент инициализации сервера.

`init()` функции могут быть полезны для статической инициализаци конфигурации на уровне пакета, например. Но не стоит их использовать для создания инстансов структур и т.д. так как это может затруднить и отладку и тестирование.

Предлагаю просто вызывать в момент инициализации приложения `NewMemStorage()` и передавать инстанс в хэндлер.

Это пока не обязательно, но в агента реализована синхронизация, а здесь нет. Предлагаю подумать нужна ли она здесь?

Fairness2 replied:
Метод init больше не используется, NewMemStorage отдаёт новое хранилище для использования вышестоящими функциями

vokinneberg commented on internal/metrics/MemStorage.go:
```diff
@@ -0,0 +1,82 @@
+	storage.metrics[name] = value
+}*/
+
+func (storage *MemStorage) SetGauge(name string, value Gauge) {
```
Названия методов не совсем консистентные. Add -> Set - думаю стоит использовать либо то, либо другое.

Fairness2 replied:
Я исходил из того, что Gauge мы перезаписываем, поэтому set, а Counter мы добавляем к существующему, поэтому Add. На мой вгляд, так понятнее, что происходит со значениями

vokinneberg replied:
Справедливо.

vokinneberg commented on internal/helpers/helpers_test.go:
```diff
@@ -0,0 +1,59 @@
+
+			assert.Equal(t, c.mockStatus, result.StatusCode)
+
+			defer result.Body.Close()
```
Стоит закрывать тело ответа сразу после полкчения, так как если assert остановит тест он так и останется не закрытым.
```suggestion
			result := response.Result()
			defer result.Body.Close()
			
			assert.Equal(t, c.mockStatus, result.StatusCode)
```

Fairness2 replied:
Перенёс функцию выше, сразу после получения

vokinneberg commented on cmd/skill/main_test.go:
```diff
@@ -0,0 +1,58 @@
+		expectedCode int
+		expectedBody string
+	}{
+		{method: http.MethodGet, expectedCode: http.StatusMethodNotAllowed, expectedBody: ""},
```
Рекомендую добавить более информативные наименования для тестов.

Fairness2 replied:
Добавил названия для тестов

vokinneberg commented on cmd/server/handlers/getmetrics/handler.go:
```diff
@@ -0,0 +1,71 @@
+
+var t map[string]*template.Template // Массив для хранения шиблонов с их именами
+
+func init() {
```
Пожалуй единственное место, где оправдано использование `init()`, так как лучше чтобы шаблон был скомпилирован единожды. 
Правда не совсем понимаю назначение использования мапы, так как тут используется только один шаблон?

Fairness2 replied:
На случай, если ещё шаблоны понадобятся, но в данном случае это излишне, согласен

vokinneberg commented on cmd/server/handlers/getmetrics/handler_test.txt:
```diff
@@ -0,0 +1,130 @@
+package getmetrics
```
Мне кажется тут не корректное расширение у файла?

Fairness2 replied:
Я написал тест, но из-за того, что при инициализации темплейтов используются относительные пути, то при тестах в гите тест падал, так как он запускается используя корневую папку - папку с тестом, а не папку проекта. Решил оставить решение на потом, и изменил расширение файла, чтобы он не попал в компиляцию

vokinneberg commented on cmd/skill/main.go:
```diff
@@ -0,0 +1,59 @@
+
+func webhook(response http.ResponseWriter, request *http.Request) {
+	// Разрешаем только POST запросы
+	if request.Method != http.MethodPost {
```
Не совсем понятно зачем нужен этот вебхук - chi должен обработать такую ситуацию автоматически.

Fairness2 replied:
Это в качестве примера в уроке по построению умения для алисы было, а так как тут не используется chi, то использование для ограничения метода оправдано🤔

vokinneberg replied:
А как вообще это используется? Это какой-то отдельный бинарник? А для чего он нужен?

Fairness2 replied:
В уроках в конце каждой главы перед инкрементом даётся пример по пройденному материалу на примере разработки навыка для алисы. Считается отдельным третьим приложением после агента и сервера

vokinneberg reviewed (changes requested):
Поздравляю с завершением первого спринта и первым ревью! 🤗 Твой код хорошо структурирован, покрыт тестами, работает и проходит тесты. Отличная работа! 🚀

**Сильные стороны твоего PR:**

- **Функциональность:**Код функционален и корректен! 🚀
- **Тестирование:**Реализованы тесты базовых сценариев и дополнительные тесты. В тестах есть развернутые сообщения о том, что пошло не так. 👍
- **Структура сервиса:**По условию задания структура сервиса должна представлять собой набор пакетов и структур, разделённых на слои. Код хорошо структурирован и разбит на отдельные компоненты и пакеты. 👍
- **Реализация интерфейсов:** Слой`storage`должен имплементировать интерфейс хранения, например`repositories`. Ты реализовал интерфейс, но я посоветовал бы определить его там, где он используется, а не рядом с реализацией.

**На что я бы рекомендовал обратить внимание и что можно улучшить:**

- **Сложность:** Мне показалось, что в некоторых местах твой код сложнее, чем мог бы быть. Например, использования рефлексии для получения имен метрик я бы избегал.
- **Глобальные переменные и метод**`init()`: Не стоит использовать инициализацию при помощи метода init(), если это явно не требуется, как в случае с темплейтами.
- **Синхронизация доступак хранилищу**. Ты реализовал синхронизацию доступа к хранилищу клиента, но не сервера. Стоит подумать, как будет себя вести код при конкурентной обработке нескольких запросов на сервере и как можно синхронизировать доступ к хранилищу.
- **Логирование:** Я бы советовал использовать либо логгер из пакета`log`, либо из пакета`slog`для структурного логирования.
- **Комментарии:** Было бы здорово, если бы ты добавил комментарии в коде там, где есть сложные места, которые требуют пояснения, а также для публичных методов.

Fairness2 committed:
iter5.1 | Код был изменён в соответствии с замечаниями из код ревью

Были проведены значительные изменения в коде, оптимизированное его работа. Многие неиспользуемые строки были удалены, а также проведена оптимизация хранилища метрик и сервера приложения. Было изменено несколько имен и других деталей для лучшей читаемости и понимания кода.

Список измененных файлов:
- cmd/agent/collector/main.go
- cmd/server/handlers/getmetric/handler_test.go
- cmd/agent/main.go
- cmd/server/handlers/getmetrics/handler_test.txt
- cmd/server/main.go
- internal/helpers/helpers_test.go
- internal/metrics/MemStorage.go
- cmd/skill/main_test.go
- cmd/server/config/config.go
- cmd/server/handlers/handlemetric/handlemetric_test.go
- internal/metrics/MemStorage_test.go
- cmd/agent/config/env.go

Fairness2 committed:
iter5.2 | Изменён URL сервера по умолчанию в коде

Список измененных файлов:
- cmd/server/config/config.go

Fairness2 committed:
iter5.3 | Упрощена система хранения и изменены параметры конфига

Типы изменены с time.Duration на int для возможности указать только число, без единицы измерения. Вместо использования функций SetGlobal для установки глобального хранилища метрик и коллекции, теперь используются напрямую.

Измененные файлы:
- cmd/agent/main.go
- cmd/server/main.go
- cmd/agent/collector/collection/initializer.go
- cmd/server/handlers/handlemetric/handlemetric_test.go
- internal/metrics/storage.go
- cmd/agent/config/parse.go

vokinneberg commented on cmd/agent/collector/collection/collection.go:
```diff
@@ -0,0 +1,77 @@
+
+	// mutex Мьютекс для устранения состояния гонки при параллельных внесениях изменений в кэш
+	// mutex TODO Пока убрал RWMutex, так как с обычным работать проще, а в команде на отправку сбрасываем каунтер
+	mutex *sync.Mutex
```
Получается в твоем случае доступ на запись  к коллекции будет блокироваться при каждом чтении. Это конечно будет не заметно при запуске тестов, но в случае более-менее серьезной нагрузки это станет проблемой. Думаю можно оставить этот момент для обсуждения до того времени, когда на курсе будет затрагиваться тема конкурентности/многопоточности.

Fairness2 replied:
Оставил так в агенте, так как у нас при чтении всё равно идёт запись в коллекцию, мы сбрасываем каунтер сбора метрик. Так как агент у нас пока не запускает кучу потоков по сбору, а только 2 потока, один для отправки, второй для сбора, то блокировка не должна приносить много проблем. В дальнейшем можно будет вернуться к этому в теме конкурентности/многопоточности.

vokinneberg commented on cmd/agent/config/env.go:
```diff
@@ -0,0 +1,50 @@
+var Params *CliConfig
+
+// InitializeNewCliConfig инициализация конфигурации приложения
+func InitializeNewCliConfig() *CliConfig {
```
Возможно стоит переименовать в `InitializeDefaultConfig`, мне кажется это будет больше соответствовать назначение функции?

Fairness2 replied:
Переименовал в InitializeDefaultConfig (будет видно в инкременте следующего спринта). Изначальное название было оправданным, когда мы создавали конфиг только из аргументов командной строки, с применением окружения название потеряло свой первоначальный смысл

vokinneberg commented on cmd/agent/config/env.go:
```diff
@@ -0,0 +1,50 @@
+}
+
+// Params конфигурация приложения
+var Params *CliConfig
```
Это безусловно твое решение использовать глобальную переменную для хранения конфигурации. По моему опыту с глобальными переменными могут возникать различные неприятные ситуации, но в нашем случае я думаю это не страшно. Тем не менее на будущее я бы рекомендовал попробовать упростить работу с конфигом - мне показалось что реализация немного переусложнена.

Fairness2 replied:
Тоже думаю, что глобальные переменные не хорошо. Но пока не придумал как сделать лучше без значительного усложнения. 
Может как-то использовать контекст? 
Конфигурация всего приложения, в Java и PHP такая конфигурация тоже получается их определённого глобального контекста🤔

vokinneberg commented on cmd/agent/collector/collection/initializer.go:
```diff
@@ -0,0 +1,44 @@
+// CollectProcess continuously collects system metrics by reading memory stats,
+// collecting the stats using collection.Collection.Collect, and then sleeping
+// for the duration defined by config.PollInterval.
+func CollectProcess(ctx context.Context) {
```
👍 за использование контекста для управления жизненным циклом функции!

vokinneberg commented on cmd/agent/collector/collection/initializer.go:
```diff
@@ -0,0 +1,44 @@
+// for the duration defined by config.PollInterval.
+func CollectProcess(ctx context.Context) {
+	log.Printf("Collect metrics process starts. Period is %d mseconds\n", config.Params.PollInterval)
+	for {
+		stats := runtime.MemStats{}
+		runtime.ReadMemStats(&stats)
+		Collection.Collect(stats)
+		// Ловим закрытие контекста, чтобы завершить обработку
+		select {
+		case <-ctx.Done():
+			log.Println("Collect metrics process stopped")
+			return
+		default:
+			//continue
+		}
+		time.Sleep(config.Params.PollInterval)
+	}
```
Тут можно немного оптимизировать, предлагаю рассмотреть такой вариант:

```suggestion
        for {
		// Ловим закрытие контекста, чтобы завершить обработку
		select {
		case <-time.After(config.Params.PollInterval):
			stats := runtime.MemStats{}
			runtime.ReadMemStats(&stats)
			Collection.Collect(stats)
		case <-ctx.Done():
			log.Printf("Collect metrics process stopped: %v\n", ctx.Err())
			return
		}
	}
```

Fairness2 replied:
Переделал с time.After (изменение будет в мерже следующего спринта). Выглядит дейтсвительно лучше. Использовал слип, так как его пока рекомендовали в задании по инкременту

vokinneberg commented on cmd/agent/main.go:
```diff
@@ -1,3 +1,58 @@
+	// Создаём новую коллекцию метрик и устанавливаем её глобально
+	collection.Collection = collection.NewCollection()
+
+	ctx, cancel := context.WithTimeout(context.Background(), 30*time.Second) // Таймер, чтобы ограничить время работы агента, чтобы пройти тесты
+	defer cancel()
+	// Создаём группу ожидания на 2 потока: сборки данных и отправки
+	wg := sync.WaitGroup{}
+	wg.Add(2)
+	go func() {
+		collection.CollectProcess(ctx)
+		defer wg.Done()
+	}() // Запускаем сборку данных
+
+	// Запускаем отправку данных
+	client := sender.NewSender(collection.Collection)
+	log.Println("New sender client created")
+	go func() {
+		client.PeriodicSender(ctx)
+		defer wg.Done()
+	}()
+
+	// Бесконечный цикл со считыванием ввода консоли, чтобы программа работала пока нужно
+	/*for {
+		fmt.Println("Agent is running. Print C to finish agent")
+		var command string
+		_, err := fmt.Fscan(os.Stdin, &command)
+		if err != nil {
+			log.Println(err)
+		}
+		if command == "C" {
+			cancel()
+			break
+		}
+	}*/
+	wg.Wait() // Ожидаем завершения всех горутин
```
Ты правильно разобрался с `WaitGroup`, это очень здорово! 👍 Могу предложить небольшое улучшение, которое поможет решить несколько проблем. Во-первых ты создал контекст с таймаутом и соотвественно он отмениться через 30 секунд и агент прекратит свою работу, а нам нужно чтобы он работал пока мы его не решили завершить принудительно. Я реализовал завершения по нажатию Ctrl+C в консоли или по отправке `SIGTERM`. 

```suggestion
ctx, cancel := context.WithCancel(context.Background()) // Таймер, чтобы ограничить время работы агента, чтобы пройти тесты
	defer cancel()

	var wg sync.WaitGroup
	wg.Add(2)

	// Создаём новую коллекцию метрик и устанавливаем её глобально
	collection.Collection = collection.NewCollection()
	go func() {
		defer wg.Done()
		collection.CollectProcess(ctx)
	}()

	// Запускаем отправку данных
	client := sender.NewSender(collection.Collection)
	go func() {
		defer wg.Done()
		client.PeriodicSender(ctx)
	}()

	// Ожидаем сигнала завершения Ctrl+C, чтобы корректно завершить работу
	stop := make(chan os.Signal, 1)
	signal.Notify(stop, syscall.SIGINT, syscall.SIGTERM)
	<-stop

	// Останавливаем все горутины
	log.Println("Agent is stopping")
	cancel()

	// Ждём завершения всех горутин
	wg.Wait()
```

Fairness2 replied:
Пока оставил таймаут, но добавил выход по Ctrl+C. Получилось так, что либо заканчивается таймер, либо закрывается приложение
```
ctx, cancel := context.WithTimeout(context.Background(), 30*time.Second) // Таймер, чтобы ограничить время работы агента, чтобы пройти тесты
	defer cancel()
	// Создаём группу ожидания на 2 потока: сборки данных и отправки
	wg := sync.WaitGroup{}
	wg.Add(2)
	go func() {
		collection.CollectProcess(ctx)
		defer wg.Done()
	}() // Запускаем сборку данных
	// Запускаем отправку данных
	client := sender.NewSender(collection.Collection)
	log.Println("New sender client created")
	go func() {
		client.PeriodicSender(ctx)
		defer wg.Done()
	}()
	// Ожидаем сигнала завершения Ctrl+C, чтобы корректно завершить работу
	stop := make(chan os.Signal, 1)
	signal.Notify(stop, syscall.SIGINT, syscall.SIGTERM)
	select {
	case <-stop:
		cancel()
	case <-ctx.Done():
		// continue
	}
	log.Println("Agent is stopping")
	wg.Wait() // Ожидаем завершения всех горутин
	log.Println("Agent stopped")
```
Будет в мерже следующего спринта

vokinneberg reviewed (approved):
Костя, привет! 👋 Мне показалось, что ты не учёл все мои замечания. Действительно, большинство из них имеют рекомендательный характер, но было бы здорово, если бы ты оставил свои комментарии к ним 😊 Здорово, что ты разобрался с синхронизацией и задумался о завершении приложения. Я добавил несколько комментариев с предложениями о том, как можно улучшить механизмы, которые ты добавил 👍 Тем не менее, мне всё ещё кажется, что с твоей реализацией можно немного поработать и сделать несколько улучшений или хотя бы обсудить их. Обязательно приходи на встречу 1 на 1 в следующем спринте, и мы бы обсудили и те комментарии, на которые ты не ответил, и новые комментарии тоже 🙌

Поздравляю с завершением первого спринта и завершением первого ревью! 🤗 Отличная работа! 🚀 И удачи на втором спринте 💪
